Neos.Media/Classes/Domain/Repository/ThumbnailRepository.php

```diff
@@ -124,4 +134,68 @@ public function findOneByAssetAndThumbnailConfiguration(AssetInterface $asset, T
         $query->setMaxResults(1);
         return $query->getOneOrNullResult();
     }
+
+    /**
+     * store via DBAL to avoid errors caused by concurrent creation of thumbnails - see
+     * https://github.com/neos/neos-development-collection/issues/3479#issuecomment-1016375400
+     *
+     * @param Thumbnail $thumbnail
+     * @param ThumbnailConfiguration $configuration
+     * @return Thumbnail|null
+     * @throws DBALException
+     * @throws \Doctrine\DBAL\Driver\Exception
+     * @throws \Doctrine\ORM\NonUniqueResultException
+     * @throws \Neos\Flow\Persistence\Exception\IllegalObjectTypeException
+     */
+    public function persistThumbnailDirectly(Thumbnail $thumbnail, ThumbnailConfiguration $configuration)
+    {
+        $thumbnailIdentifier = $this->persistenceManager->getIdentifierByObject($thumbnail);
+        $assetIdentifier = $this->persistenceManager->getIdentifierByObject($thumbnail->getOriginalAsset());
+        $thumbnailResource = $thumbnail->getResource();
+
+        $sql = 'INSERT INTO neos_media_domain_model_thumbnail (persistence_object_identifier, originalasset, resource, width, height, configuration, configurationhash, staticresource, quality) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)';
+        $params = [
+            $thumbnailIdentifier,
+            $assetIdentifier,
+            null,
+            $thumbnail->getWidth(),
+            $thumbnail->getHeight(),
+            json_encode($configuration->toArray()),
+            $configuration->getHash(),
+            $thumbnail->getStaticResource(),
+            $thumbnail->getQuality(),
+        ];
+
+        $connection = $this->entityManager->getConnection();
+        try {
+            $affectedRows = $connection->prepare($sql)->executeStatement($params);
+            if ($affectedRows === 0) {
+                $this->logger->error('Could not persist thumbnail', LogEnvironment::fromMethodName(__METHOD__));
+            }
+        } catch (DBALException\UniqueConstraintViolationException $e) {
+            // ignore, the thumbnail has been generated already; re-fetch
+            $this->logger->debug('Persisting thumbnail caused a UniqueConstraintViolationException, ignoring: ' . $e->getMessage(), LogEnvironment::fromMethodName(__METHOD__));
+        } catch (DBALException\ForeignKeyConstraintViolationException $e) {
+            // TODO
+            // why do we end up here, this method is called only if !$this->persistenceManager->isNewObject($asset)
+            // but actually that check doesn't even seem to make a difference…
+            $this->logger->debug('Persisting thumbnail caused a ForeignKeyConstraintViolationException, ignoring: ' . $e->getMessage(), LogEnvironment::fromMethodName(__METHOD__));
+            // ignore, the asset has not been persisted yet, rely on "cascade" persistence later
+            return $thumbnail;
```

Comment:
Well… I have my doubts here. This code should not be run, if the asset is "new" (see check in ThumbnailService), but still is run, e.g. when uploading a new asset in the media browser.